File: app/javascript/channels/ambulance_location_channel.js
// import consumer from "channels/consumer"

// const ambulanceId = document.getElementById("ambulance-id").value;

// if (ambulanceId) {
//   console.log("Ambulance ID:", ambulanceId);
//   console.log("Subscribing to:", { channel: "AmbulanceLocationChannel", ambulance_id: ambulanceId });

//   consumer.subscriptions.create(
//     { channel: "AmbulanceLocationChannel", ambulance_id: ambulanceId },
//     {
//       connected() {
//         console.log("Connected to WebSocket");
//       },
//       disconnected() {
//         console.log("Disconnected from WebSocket");
//       },
//       received(data) {
//         console.log("Received data:", data); // Confirm the data is received
//         const latitudeInput = document.getElementById('latitude');
//         const longitudeInput = document.getElementById('longitude');

//         if (latitudeInput && longitudeInput) {
//           latitudeInput.value = data.latitude;
//           longitudeInput.value = data.longitude;
//         }
//       }
//     }
//   );
// } else {
//   console.error("Ambulance ID is missing.");
// }

import consumer from "channels/consumer"

const ambulanceId = document.getElementById("ambulance-id").value;

let ambulanceLocationChannel;

export function getAmbulanceLocationChannel(ambulanceId) {
  if (ambulanceLocationChannel) {
    return ambulanceLocationChannel;
  }

  ambulanceLocationChannel = consumer.subscriptions.create(
    { channel: "AmbulanceLocationChannel", ambulance_id: ambulanceId },
    {
      connected() {
        console.log(`Connected to WebSocket for Ambulance ${ambulanceId}`);
      },

      disconnected() {
        console.log("Disconnected from WebSocket");
      },

      received(data) {
        console.log("Received data:", data);
      },

      updateLocation(latitude, longitude) {
        console.log(`Sending latitude: ${latitude}, longitude: ${longitude}`);
        this.perform("update_location", { latitude, longitude });
      }
    }
  );

  return ambulanceLocationChannel;
}

File: app/channels/ambulance_location_channel.rb
class AmbulanceLocationChannel < ApplicationCable::Channel
  def subscribed
    # Stream specific to the ambulance being tracked
    Rails.logger.debug("Params: #{params.inspect}") # Log all params
    Rails.logger.debug("Subscribing to ambulance_location_#{params[:ambulance_id]}")

    # Ensure ambulance_id is being passed and is not nil
    if params[:ambulance_id].present?
      stream_from "ambulance_location_#{params[:ambulance_id]}"
    else
      Rails.logger.debug("ambulance_id is nil")
    end
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
    stop_all_streams
  end

  def receive(data)
    Rails.logger.debug("Received data: #{data.inspect}")
  
    if data['ambulance_id'].present?
      ActionCable.server.broadcast("ambulance_location_#{data['ambulance_id']}", {
        latitude: data['latitude'],
        longitude: data['longitude']
      })
    else
      Rails.logger.debug("ambulance_id is missing in the received data")
    end
  end
end
